File: backend/app/src/main/kotlin/net/animeta/backend/controller/v2/UserPostsController.kt
package net.animeta.backend.controller.v2

import net.animeta.backend.db.Datastore
import net.animeta.backend.db.query
import net.animeta.backend.dto.PostDTO
import net.animeta.backend.exception.ApiException
import net.animeta.backend.model.QHistory.history
import net.animeta.backend.repository.UserRepository
import net.animeta.backend.serializer.PostSerializer
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/v2/users/{name}/posts")
class UserPostsController(val userRepository: UserRepository,
                          val datastore: Datastore,
                          val postSerializer: PostSerializer) {
    @GetMapping
    fun get(@PathVariable name: String,
            @RequestParam("before_id", required = false) beforeId: Int?,
            @RequestParam(defaultValue = "32") count: Int,
            @RequestParam(required = false) options: PostSerializer.Options?): List<PostDTO> {
        val user = userRepository.findByUsername(name) ?: throw ApiException.notFound()
        var query = history.query.selectRelated(history.record)
                .filter(history.user.eq(user))
                .orderBy(history.id.desc())
        if (beforeId != null) {
            query = query.filter(history.id.lt(beforeId))
        }
        val limit = minOf(count, 128)
        return datastore.query(query.limit(limit))
                .map { postSerializer.serialize(it, options ?: PostSerializer.legacyOptions(includeRecord = true)) }
    }
}

File: backend-ts/src/controllers/user_posts.controller.ts
import { Controller, DefaultValuePipe, Get, Param, Query } from "@nestjs/common";
import { PostDTO } from 'shared/types';
import { ApiException } from "./exceptions";
import { LessThan, Repository } from "typeorm";
import { InjectRepository } from "@nestjs/typeorm";
import { PostSerializer, PostSerializerOptions } from "src/serializers/post.serializer";
import { History } from 'src/entities/history.entity'
import { User } from "src/entities/user.entity";
import { ParseIntOrNullPipe } from "src/utils/parse_int_or_null.pipe";

@Controller('/api/v4/users/:name/posts')
export class UserPostsController {
  constructor(
    @InjectRepository(User) private userRepository: Repository<User>,
    @InjectRepository(History) private historyRepository: Repository<History>,
    private postSerializer: PostSerializer,
  ) {}

  @Get()
  async get(
    @Param('name') name: string,
    @Query('before_id', new ParseIntOrNullPipe()) beforeId: number | null,
    @Query('count', new ParseIntOrNullPipe(), new DefaultValuePipe(32)) count: number,
    @Query('options') optionsJson: string | null,
  ): Promise<PostDTO[]> {
    const user = await this.userRepository.findOne({ where: {username: name} })
    if (!user) throw ApiException.notFound()
    const posts = await this.historyRepository.find({
      relations: ['record'],
      where: {
        user,
        ...beforeId ? { id: LessThan(beforeId) } : {},
      },
      order: { id: 'DESC' },
      take: Math.min(count, 128),
    })
    const options: PostSerializerOptions =
      optionsJson ? JSON.parse(optionsJson) : PostSerializer.legacyOptions({
        includeRecord: true,
      })
    return await Promise.all(posts.map(it => this.postSerializer.serialize(it, options)))
  }
}
